# INTRODUCTION TO PYTHON

## Final Exam 
## Total Marks : 70 Marks

----

# SECTION A: 20 MARKS

 _________________

### Q1.  Write a program that asks the user to enter size of the list, then do the following: (5Marks)
(a) if size of the list is less than 4 then print a message to user "Invalid size and enter a size more than 4". The program will continue from begining.

(b) After taking valid size , program should take list items from user and create user defined list. 

(c) Use the choice() function to print out random item from the list.

(d) Use the sample() function to print out three random items from the list.

(e) Use the shuffle() function to reorder the list and then print out the list.


**Choice(),shuffle(),sample() functions are defined under random library. You need to import random library to use it.**

***
#### Sample Input/Output:


<span style="font-family:consolas ">

Enter number of elements in the list: 3

Invalid size and enter a size more than 4

Enter number of elements in the list: 5

1

2

3

4

5

Random item from the list:  1

Three random items from the list: ['1', '4', '2']

Shuffeled list: ['3', '4', '5', '1', '2']

***

In [28]:
def checkSize():
    size = int(input("Enter the size of the list : "))
    if size < 4:
        print("Invalid size and enter a size more than 4")
        checkSize()
    else:
        myList=[i  for i in range(1,size+1)]
        print(myList)
        a=rd.choice(myList)
        print("Random item from the list : ",a)
        x=rd.sample(myList,3)
        print("Three random items from the list : ",x)
        y=rd.shuffle(myList)
        print("Shuffeled list : ",y)

In [9]:
checkSize()

Enter the size of the list : 5
[1, 2, 3, 4, 5]
Random item from the list :  1
Three random items from the list :  [3, 1, 5]
Shuffeled list :  None


In [2]:
import random as rd;

### Q2. Ask the user for a number and then print the following, where the pattern ends at the number that the user enters.
<pre>
1
 2
  3
   4
    5
   4
  3
 2
1
</pre>

***
#### Sample Input/Output:

***

In [11]:
n=int(input("Enter a number : "))
for i in range(1,n+1):
    print(' '*(i-1),i)
for j in range(n-1,0,-1):
    print(' '*(j-1),j)
    

Enter a number : 6
 1
  2
   3
    4
     5
      6
     5
    4
   3
  2
 1


### Q3.Write a program to validate the data entered by its user and then perform conversion from kilogram to pounds. Write a program that asks the user for a weight and converts it from kilograms to pounds. Whenever the user enters a weight below 0, the program should tell them that their entry is invalid and then ask them again to enter a weight. At the end it should show a weight converted into pounds. do this repeatedly untill user enters 'STOP' [use of IF statement is restricated here] (5 marks)


***
#### Sample Input/Output:

Enter Weight or enter STOP: 10

Weight in pounds:  22.046226218 lbs

Enter Weight: 50

Weight in pounds:  110.23113109 lbs

Enter Weight: 45

Weight in pounds:  99.208017981 lbs

Enter Weight: 62

Weight in pounds:  136.6866025516 lbs

Enter Weight: STOP

***

In [24]:
def func():
    n=input("Enter Weight or enter STOP : ")
    if n.upper()=='STOP':
        print()
    else:
        if int(n) < 0:
            print("Entry is invalid")
            func()
        else:
            x = int(n)*2.20462
            print("Weight in pounds : ",x," lbs")
            func()

In [25]:
func()

Enter Weight or enter STOP : 10
Weight in pounds :  22.0462  lbs
Enter Weight or enter STOP : 50
Weight in pounds :  110.231  lbs
Enter Weight or enter STOP : 45
Weight in pounds :  99.2079  lbs
Enter Weight or enter STOP : 62
Weight in pounds :  136.68643999999998  lbs
Enter Weight or enter STOP : STOP



### Q4. Ask the user to enter a list of 10 elemnts containing numbers between 1 and 15. Then replace all of the entries in the list that are greater than 10, with 10 using map() function.(5 Marks)


***
#### Sample Input/Output:

Enter a number between 1 to 15: 3

Enter a number between 1 to 15: 13

Enter a number between 1 to 15: 7

Enter a number between 1 to 15: 12

Enter a number between 1 to 15: 13

Enter a number between 1 to 15: 4

Enter a number between 1 to 15: 5

Enter a number between 1 to 15: 6

Enter a number between 1 to 15: 17

Enter a number between 1 to 15: 18

Output list :  [3, 10, 7, 10, 10, 4, 5, 6, 10, 10]

***

In [27]:
list1=[]
for i in range(1,11):
    n = int(input("Enter a number between 1 to 15: "))
    list1.append(n)
list2=list(map(lambda x:10 if x>10 else x,list1))
print(list2)

Enter a number between 1 to 15: 3
Enter a number between 1 to 15: 13
Enter a number between 1 to 15: 7
Enter a number between 1 to 15: 12
Enter a number between 1 to 15: 13
Enter a number between 1 to 15: 4
Enter a number between 1 to 15: 5
Enter a number between 1 to 15: 6
Enter a number between 1 to 15: 17
Enter a number between 1 to 15: 18
[3, 10, 7, 10, 10, 4, 5, 6, 10, 10]


# SECTION B: 20 MARKS

---------------------

### Q5. A superstore has multiple items to sell. These items are categorised as  category A and category B. For festival time the superstore decided to declare some discount, as given below. 

### for category A items, it charges Rs.15 per item if you buy less than 30 quantity of an item. If you buy between 30 and 99 quantity of an item, the cost is Rs.12 per item. If buying quantity is 100 or more for an item, the cost is Rs.10 per item. 

### For category B items, it charges Rs.30 per item if you buy less than 50 quantity of an item. If you buy between 50 and 99 quantity of an item, the cost is Rs.25 per item. If buying quantity is 100 or more for an item, the cost is Rs.20 per item. Further it also gives discount of 5% if total bill exceeds 1000 and 10% if bill exceeds 1500.

### Write a program which repeatedly ask user to enter no of items and its category. If user enters 'STOP' it shows user has done with the purchase, now calculate total payable amount and display it to user(10 Marks)


***
#### Sample Input/Output:
Enter no of items and its category: 23,A

Enter no of items and its category: 56, b

Enter no of items and its category: 67,b

Enter no of items and its category: 110,a

Enter no of items and its category: STOP

Total amount payable is :  2808.0

***

### Q6.Create a 5*5  list of numbers from 1 to 10 using list comprehension. Print the nested list. 
### Create a dictionary whose keys are the numbers and whose values are the frequencies of the number in the nested list mentined above
### Print the dictionary in the descending order of frequencies of the numbers. 
### Then print the three most common numbers.(10 Marks) 


***
#### Sample Input/Output:

5 * 5 nested list : 
[[10, 3, 10, 9, 4], [1, 6, 1, 2, 5], [4, 1, 5, 9, 7], [9, 7, 5, 4, 1], [8, 7, 3, 3, 3]]


Dictionary for number and its frequency : 
{1: 4, 2: 1, 3: 4, 4: 3, 5: 3, 6: 1, 7: 3, 8: 1, 9: 3, 10: 2}


Dictionary in descending order of frequency : 
{1: 4, 3: 4, 4: 3, 5: 3, 7: 3, 9: 3, 10: 2, 2: 1, 6: 1, 8: 1}


Top three frequent numbers are [1, 3, 4]

***

# SECTION C: 30 MARKS

--------------------

### Q7. One way of encrypting a string message is to rearrange its characters. Simple way to rearrange the characters is to pick out the characters at even indices, put them first in the encrypted string and follow them by the odd characters.
#### For example, the string message would be encrypted as msaeesg because the even characters are m, s, a, e (at indices 0, 2, 4, and 6) and the odd characters are e, s, g (at indices 1, 3, and 5).
#### (a) Write a function encrypt()to encrypt the string given by user.
#### (b) Write a function decrypt() to decrypt a string that was encrypted before.(15 Marks)


***
#### Sample Input/Output:
<span style="font-family:consolas ">
Encrypted message is :  SceMsaeertesg
    
Decrypted message is :  SecretMessage
***

### Q8. Write a program that checks to see if a 4 * 4 list is a magic square. In a magic square, every row, column, and the two diagonals add up to the same value. Write a function magic_square which returns True if list is magic square otherwise it returns false (15 Marks)

E.g Given below is 4 * 4 magic Square
<pre>
    1  15  14  4
    10 11   8  5
    7  6    9  12
    16 2    3  13


***
#### Sample Input/Output: 
Its a magic Square

***